AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java

```
@@ -336,23 +335,74 @@ public static void deleteACRALimiterData(Context context) {
         context.getFileStreamPath("ACRA-limiter.json").delete();
     }
 
+    /**
+     *  Returns a Context with the correct, saved language, to be attached using attachBase().
+     *  For old APIs directly sets language using deprecated functions
+     *
+     * @param remoteContext The base context offered by attachBase() to be passed to super.attachBase().
+     *                      Can be modified here to set correct GUI language.
+     */
+    @SuppressWarnings("deprecation")
+    public static Context updateContextWithLanguage(Context remoteContext){
+        SharedPreferences preferences;
```

My main concern is error handling here, could this method be wrapped in a try..catch..send error report and return the original context if this fails.

---

Can do that, but I'm unsure, which exceptions we should catch? All? This is (as far as I looked up) considered bad practice. And as long as we don't know, which exceptions might occur, we can't narrow it down.

Do you have a specific reason why you want a try-catch block here? Do you want it as a temporary measure, so when things get rolled out and it fails for 1% or the users, the impact gets reduced and the block gets removed later on, when no crashes occur?

If you want a exception handler here, I can add it.

---

TL;DR: Yes, but discussion is valuable, this is my opinion, and disagreements are more than welcome

* Java has a distinction between `Exception` and `Error`. Typically, you shouldn't try to catch an `Error` (OutOfMemoryError for example). 
* You shouldn't catch and ignore exceptions that you can't resolve, so somewhere higher up the stack will handle them
* You should handle exceptions individually, as you have the most context as to what's gone wrong and can take specific corrective action
* In this case, we can't easily anticipate all of the exceptions that will be thrown, and the list of exceptions that may be thrown is large.
* It's unlikely that any of the exceptions that we catch have a specific corrective action to be taken
* Throwing would crash the app, which is a serious issue if it occurs on startup. Even more serious as the base context may be attached before our error handler is initialised

----

In our case:

* We're at the point in the stack where we have a reasonable default to return on any exception
* The catch block is very unlikely to be executed
* If the catch block is executed, we call `AnkiDroidApp.sendExceptionReport(e, "updateContextWithLanguage");`
  * This allows us to diagnose the crash, the user will be notified of the problem, or the app will work as normal for them
  * This will do nothing before ACRA is set up, but we'll get an exception report on the second call
* It's a lot of effort to enumerate the possible exceptions which can be thrown
  * Even if we did, it's unlikely that any of them have a reasonable corrective action to take over the generic: "Return the `context` that we were provided with".
* It's reasonable for the contract of the method to state that the input will be returned if there is a failure.


----

Counterpoints:

* A `catch Exception` will catch `DeadSystemException`, but we're dead in the water anyway if this occurs.

---

Catching Exception is not a bad practice, you are reading information from people that have studied a lot I think, but not programmed in the real world.

If you don't catch Exception the app will crash here, which is terrible. It's as simple as that.

Further, as David mentioned, no one else can handle any errors coming out of here better than we can, so you catch everything and return a reasonable default.

Programming is pragmatic not dogmatic. Catching Exception is fine. Catching Error *sometimes* is fine. Exponential cost algorithms are sometimes fine. Not failing fast is sometimes fine. It all depends on the cost/benefit in the area you are programming in.